### PR TITLE
Added patterns for randperm + index_add

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112102
* #112101
* #112093

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
Noice. Do you have CPU/CUDA benchmarks to make sure that these optimisations are fusing as intended and we do get a speed-up? Either that or a test introspectingt he CPU/CUDA code would be nice.

**Comment by Chillee** [comment]
@lezcano Of course I have benchmarks :P

Without this optimization
```
eager fwd: 1159.25 us
compiled fwd: 463.3 us
eager fwd + bwd: 2289.7 us
compiled fwd + bwd: 1103.97 us
```

With this optimization
```
eager fwd: 1157.48 us
compiled fwd: 408.56 us
eager fwd + bwd: 2288.23 us
compiled fwd + bwd: 1002.37 us
```

**Comment by fmassa** [comment]
This is great @Chillee !

Would it be possible for a 3rd-party developer to write similar plugins for inductor? And if yes, what is the best way of doing it?

**Review by lezcano — APPROVED** [review]

File: torch/_inductor/fx_passes/misc_patterns.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    def randperm_index_replacement(x, slice_shape):
+        index = torch.randperm(x.shape[0], device=x.device)[:slice_shape]
+        return torch.ops.aten._unsafe_index(x, (index,)), index
+
+    pattern = register_replacement(
+        randperm_index_pattern,
+        randperm_index_replacement,
+        [torch.empty(4, 8, device=device)],
+        fwd_only,
+        [post_grad_patterns, joint_graph_patterns],
```
Comment (eellison) [inline torch/_inductor/fx_passes/misc_patterns.py:62]: How does this work with joint_graph but fwd_only ?

Comment (Chillee) [reply torch/_inductor/fx_passes/misc_patterns.py:62]: It’s only the fwd graph but the pattern can show up anywhere, including the joint graph.

**Comment by Chillee** [comment]
@fmassa Yeah definitely, I think we should try and expose pattern-matching to end users.

Today, we already allow people to interpose some custom passes using `torch._inductor.config.post_grad_custom_post_pass`, but we don't have a pattern-matcher type API exposed, which is perhaps easier to use.

There are some rough spots though (like dealing with scalar values), might need to figure those out first.

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/config.py#L68

**Comment by Chillee** [comment]
Landed in https://github.com/pytorch/pytorch/pull/112476

**Pull request closed without merging**